### DISCUSSION TITLE
some experiments with libdwarf.a

### DISCUSSION THREAD

**Discussion by lonejack** [davea42/libdwarf-code#198] [Q&A] [answered]
Hi David,
I'm trying to make some experiments with your library.
I checkout the v0.8.0.
Compiled with autotools + cmake, same results, libdwarf.a created.
The configuration+compilation goes well with both methods.

Now, I created a simple main file with requested inclusion...
Here what I get:
claudio@claudio-i7:~/extension/dev/glib-test$ gcc -g -omain main.c viewer_file.o libdwarf.a `pkg-config --cflags --libs glib-2.0 gobject-2.0` -lc -DLIBDWARF_STATIC
/usr/bin/ld: libdwarf.a(libdwarf_la-dwarf_init_finish.o): in function `do_decompress':
/home/claudio/extension/dev/libdwarf-code/build/src/lib/libdwarf/../../../../src/lib/libdwarf/dwarf_init_finish.c:1387: undefined reference to `uncompress'
collect2: error: ld returned 1 exit status

I moved the libdwarf.a in the same directory where stays "main.c".
Any idea why I get the error?
Thank you

**Comment by lonejack** [comment]
This works:
 gcc -g -omain main.c viewer_file.o libdwarf.a `pkg-config --cflags --libs glib-2.0 gobject-2.0` -lc **-lz** -DLIBDWARF_STATIC

**Comment by davea42** [comment, marked as answer]
I'm feeling a bit confused because with a command line like yours I get
gcc: error: unrecognized command-line option ‘--cflags’
gcc: error: unrecognized command-line option ‘--libs’; did you mean ‘--libs=’?

Oh. This is because you lost the back-quotes, failing to mark your actual line as code to the discussion . Ok. No problem.

Question: in the place you did your libdwarf build there will be a config.h
Please do 
grep ZLIB config.h
grep ZSTD config.h

and report what these do.

What I get here is

q3 543: grep ZLIB config.h
#define HAVE_ZLIB 1
#define HAVE_ZLIB_H 1
q3 544: grep ZSTD config.h
#define HAVE_ZSTD 1
#define HAVE_ZSTD_H 1

I suspect you have just
#define HAVE_ZLIB 1
#define HAVE_ZLIB_H 1

In that case, you MUST link in -lz yourself, since libdwarf.a is an archive and 
an archive keeps no
record of what it will need to link to.

I v0.9.0  unless one has both ZLIB and ZSTD a similar situation arises
but one would need    ` -lz -lzstd` on the compile command line.
Indeed you may need both in v0.8.0 (it just depends on the
libdwarf.a  build-time config.h)

I hope this helps.
DavidA

**Reply by lonejack** [reply]
Problems on copy/paste:
gcc -g -omain main.c viewer_file.o libdwarf.a \`pkg-config --cflags --libs glib-2.0 gobject-2.0\` -lc -lz -DLIBDWARF_STATIC

claudio@claudio-i7:~/extension/dev/glib-test$ pkg-config --cflags --libs glib-2.0 gobject-2.0
-I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -lgobject-2.0 -lglib-2.0

**Reply by lonejack** [reply]
I run commands you asked. This is the result:

claudio@claudio-i7:\~/extension/dev/libdwarf-code/build$ grep ZLIB config.h
#define HAVE_ZLIB 1
#define HAVE_ZLIB_H 1
claudio@claudio-i7:\~/extension/dev/libdwarf-code/build$ grep ZSTD config.h
\/\* #undef HAVE_ZSTD \*\/
\/\* #undef HAVE_ZSTD_H \*\/
you're rigth.
